paddle/fluid/operators/prelu_op.cu

```diff
@@ -55,10 +59,116 @@ class CUDAPReluKernel : public framework::OpKernel<T> {
   }
 };
 
+template <typename T>
+struct IdentityFunctor {
+  HOSTDEVICE inline T operator()(const T& x) const { return x; }
+};
+
+template <typename T>
+__global__ void PReluGradElementWiseKernel(const T* x_ptr_, const T* y_ptr_,
+                                           const T* alpha_ptr_,
+                                           const T* dy_ptr_, T* dx_ptr_,
+                                           T* dalpha_ptr_, size_t channel,
+                                           size_t spatial_size) {
+  size_t offset = blockIdx.x * spatial_size;
+  const T* x_ptr = x_ptr_ + offset;
+  const T* y_ptr = y_ptr_ + offset;
+  const T* dy_ptr = dy_ptr_ + offset;
+  const T* alpha_ptr = y_ptr_ + offset;
+  T* dx_ptr = dx_ptr_ + offset;
+  T* dalpha_ptr = dalpha_ptr_ + offset;
+
+  for (size_t i = threadIdx.x; i < spatial_size; i += blockDim.x) {
+    T y = y_ptr[i];
+    T x = x_ptr[i];
+    T alpha = alpha_ptr[i];
+    T dy = dy_ptr[i];
+    if (dx_ptr != nullptr) dx_ptr[i] = (y > 0) ? dy : alpha * dy;
+    if (dalpha_ptr != nullptr) dalpha_ptr[i] = (x > 0) ? 0 : dy;
+  }
+}
+
+template <typename T>
+class PreluGradElementwiseFunctor {
+ public:
+  void operator()(cudaStream_t stream, const T* x, const T* y, const T* alpha,
+                  const T* dy, T* dx, T* dalpha, std::vector<int> input_shape) {
+    size_t unroll = input_shape[0] * input_shape[1];
+    size_t spatial_size = input_shape[2] * input_shape[3];
+    CHECK_LT(unroll, CUDA_MAX_NUM_BLOCKS);
+    PReluGradElementWiseKernel<T><<<unroll, CUDA_NUM_THREADS, 0, stream>>>(
+        x, y, alpha, dy, dx, dalpha, input_shape[1], spatial_size);
+  }
+};
+
+template <typename DeviceContext, typename T>
+class CUDAPReluGradKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto* x = context.Input<Tensor>("X");
+    auto* y = context.Input<Tensor>("Out");
+    auto* alpha = context.Input<Tensor>("Alpha");
+    auto* dx = context.Output<Tensor>(framework::GradVarName("X"));
+    auto* dy = context.Input<Tensor>(framework::GradVarName("Out"));
+    auto* dalpha = context.Output<Tensor>(framework::GradVarName("Alpha"));
+
+    const T* x_ptr = x->data<T>();
+    const T* y_ptr = y->data<T>();
+    const T* alpha_ptr = alpha->data<T>();
+    const T* dy_ptr = dy->data<T>();
+    T* dx_ptr = dx ? dx->mutable_data<T>(context.GetPlace()) : nullptr;
+    T* dalpha_ptr =
+        dalpha ? dalpha->mutable_data<T>(context.GetPlace()) : nullptr;
+
+    auto& mode = context.Attr<std::string>("mode");
+
+    int numel = x->numel();
+    auto dim = x->dims();
```

Comment:
Return type is DDim, don't need convert to vector

Reply:
done